src/R2DH_alg.jl

```diff
@@ -0,0 +1,283 @@
+    prox!(xk, h, x0, one(eltype(x0)))
+    hk = h(xk[selected])
+    hk < Inf || error("prox computation must be erroneous")
+    verbose > 0 && @debug "R2DH: found point where h has value" hk
+  end
+  hk == -Inf && error("nonsmooth term is not proper")
+
+  xkn = similar(xk)
+  s = zero(xk)
+  ψ = has_bnds ? shifted(h, xk, l_bound - xk, u_bound - xk, selected) : shifted(h, xk)
+
+  Fobj_hist = zeros(maxIter)
+  Hobj_hist = zeros(maxIter)
+  Complex_hist = zeros(Int, maxIter)
+  if verbose > 0
+    #! format: off
+    @info @sprintf "%6s %8s %8s %7s %8s %7s %7s %7s %1s" "iter" "f(x)" "h(x)" "√ξ" "ρ" "σ" "‖x‖" "‖s‖" ""
+    #! format: off
+  end
+
+  local ξ
+  k = 0
+  σk = sum ? σmin : max(1 / ν, σmin)
+
+  fk = f(xk)
+  ∇fk = similar(xk)
+  ∇f!(∇fk, xk)
+  ∇fk⁻ = copy(∇fk)
+  Dk = spectral ? SpectralGradient(hess_init_val, length(xk)) :
+    ((Bk isa UniformScaling) ? DiagonalQN(fill!(similar(xk), hess_init_val), psb, andrei, wolk) : DiagonalQN(diag(Bk), psb, andrei, wolk))
+  DkNorm = norm(Dk.d, Inf)
+  σkdk = sum ? Dk.d .+ σk : Dk.d  .* σk 
+  ν = 1 / σkdk[1]
+  mν∇fk = -ν * ∇fk  
+
+  optimal = false
+  tired = maxIter > 0 && k ≥ maxIter || elapsed_time > maxTime
+
+  while !(optimal || tired)
+    k = k + 1
+    elapsed_time = time() - start_time
+    Fobj_hist[k] = fk
+    Hobj_hist[k] = hk
+    σkdk = max.(σkdk, eps())
+
+    # model with diagonal hessian 
+    φ(d) = ∇fk' * d + (d' * (σkdk .* d)) / 2
+    mk(d) = φ(d) + ψ(d)
+
+    if spectral
+      prox!(s, ψ, mν∇fk, ν)
+    else
+      iprox!(s, ψ, ∇fk, σkdk)
+    end
+
+    Complex_hist[k] += 1
+    xkn .= xk .+ s
+    fkn = f(xkn)
+    hkn = h(xkn[selected])
+    hkn == -Inf && error("nonsmooth term is not proper")
+
+    Δobj = fk + hk - (fkn + hkn) + max(1, abs(fk + hk)) * 10 * eps()
+    ξ = hk - mk(s) + max(1, abs(hk)) * 10 * eps()
+
+    if ξ ≥ 0 && k == 1
+      ϵ += ϵr * sqrt(ξ)  # make stopping test absolute and relative
+    end
+    
+    if (ξ < 0 && sqrt(-ξ) ≤ neg_tol) || (ξ ≥ 0 && sqrt(ξ) < ϵ)
+        # the current xk is approximately first-order stationary
+      optimal = true
+      continue
+    end
+
+    ξ > 0 || error("R2DH: prox-gradient step should produce a decrease but ξ = $(ξ)")
+    
+    ρk = Δobj / ξ
+
+    σ_stat = (η2 ≤ ρk < Inf) ? "↘" : (ρk < η1 ? "↗" : "=")
+
+    if (verbose > 0) && (k % ptf == 0)
+      #! format: off
+      @info @sprintf "%6d %8.1e %8.1e %7.1e %8.1e %7.1e %7.1e %7.1e %1s" k fk hk sqrt(ξ) ρk σk norm(xk) norm(s) σ_stat
+      #! format: on
+    end
+
+    if η2 ≤ ρk < Inf
+      σk = max(σk / γ, σmin)
+    end
+
+    if η1 ≤ ρk < Inf
+      xk .= xkn
+      has_bnds && set_bounds!(ψ, l_bound - xk, u_bound - xk)
+      fk = fkn
+      hk = hkn
+      shift!(ψ, xk)
+      ∇f!(∇fk, xk)
+      push!(Dk, s, ∇fk - ∇fk⁻) # update QN operator
+      DkNorm = norm(Dk.d, Inf) 
+      ∇fk⁻ .= ∇fk
+    end
+
+    if ρk < η1 || ρk == Inf
+      σk = σk * γ
+    end
+
+    σkdk = sum ? Dk.d .+ σk : Dk.d  .* σk 
```

Comment:
Same here, you can use `.=` or `@.` for the whole line to avoid allocating a vector every time. If this does not work with the `? ... : ...` synthax, you can split it with a `if .. else`

Reply:
Also maybe using `sum` as a bool here is a bit ambiguous here since this is also the name of a function? Could you find another name?

Reply:
Ok thank you